### PR TITLE
feat: add slippage to atomic order buy and sell

### PR DESCRIPTION
**Pull request by Rickk137**

### CODE REVIEW

**Review by sunnyvempati — COMMENTED** [review]
can you add a specific test for this `minimumReceived` and check nothing gets deposited to market manager, nothing is collected by fee collector, everything is reverted properly?

File: markets/spot-market/test/AtomicOrderModule.buy.test.ts [hunk 1]
```diff
@@ -23,23 +23,44 @@ describe('Atomic Order Module buy()', () => {
...
+      );
+    });
+
+    it('trader1 has 0 snxETH', async () => {
+      assertBn.equal(await synth.balanceOf(await trader1.getAddress()), bn(0));
+    });
+
+    it('market withdrawable Usd has not change', async () => {
+      assertBn.equal(await systems().Core.getWithdrawableMarketUsd(marketId()), withdrawableUsd);
+    });
```
Comment (sunnyvempati) [inline markets/spot-market/test/AtomicOrderModule.buy.test.ts:54]: checking revert is prolly enough but good sanity checks 😄

**Review by sunnyvempati — APPROVED** [review]
🚢 it

File: markets/spot-market/contracts/interfaces/IAtomicOrderModule.sol [hunk 2]
```diff
@@ -53,7 +57,11 @@ interface IAtomicOrderModule {
      * @param amountUsd Amount of snxUSD trader is providing allownace to for the trade.
      * @return synthReturned Synth received on the trade based on amount provided by trader.
```
Comment (noahlitvin) [inline markets/spot-market/contracts/interfaces/IAtomicOrderModule.sol:58]: Add natspect for `minAmountReceived` plz

File: markets/spot-market/contracts/interfaces/IAtomicOrderModule.sol [hunk 3]
```diff
@@ -63,5 +71,9 @@ interface IAtomicOrderModule {
      * @param sellAmount Amount of synth trader is trading for snxUSD.
      * @return amountReturned Amount of snxUSD returned to user based on synth provided by trader.
```
Comment (noahlitvin) [inline markets/spot-market/contracts/interfaces/IAtomicOrderModule.sol:72]: Add natspect for `minAmountReceived` plz

File: markets/spot-market/contracts/modules/AtomicOrderModule.sol [hunk 4]
```diff
@@ -48,6 +52,11 @@ contract AtomicOrderModule is IAtomicOrderModule {
             amountUsable,
             SpotMarketFactory.TransactionType.BUY
         );
+
+        if (synthAmount < minAmountReceived) {
+            revert InsufficientReturnAmount(minAmountReceived, synthAmount);
```
Comment (noahlitvin) [inline markets/spot-market/contracts/modules/AtomicOrderModule.sol:57]: Rename to `InsufficientAmountReceived` ?

File: markets/spot-market/contracts/modules/AtomicOrderModule.sol [hunk 5]
```diff
@@ -99,6 +112,10 @@ contract AtomicOrderModule is IAtomicOrderModule {
             ITokenModule(spotMarketFactory.usdToken).transfer(msg.sender, returnAmount);
         }
 
+        if (returnAmount < minAmountReceived) {
+            revert InsufficientReturnAmount(minAmountReceived, returnAmount);
```
Comment (noahlitvin) [inline markets/spot-market/contracts/modules/AtomicOrderModule.sol:116]: Rename to InsufficientAmountReceived ?

Comment (noahlitvin) [reply markets/spot-market/contracts/modules/AtomicOrderModule.sol:116]: Also, maybe we move this up under the process fee call? Would save a little gas I think?

**Review by noahlitvin — CHANGES_REQUESTED** [review]

**Review by noahlitvin — APPROVED** [review]

**Pull request merged**